Updated Presentation Flow (#610)
italia/eid-wallet-it-docs: docs/en/proximity-flow.rst

        {
          "1": 2, % kty:EC2 (Elliptic curves with x and y coordinate pairs)
          "-1": 1, % crv:p256
          "-2":"5A88D182BCE5F42EFA59943F33359D2E8A968FF289D93E5FA444B624343167FE", % x-coordinate
          "-3":"B16E8CF858DDC7690407BA61D4C338237A8CFCF3DE6AA672FC60A557AA32FC67" % y-coordinate
        }
        >>)
    ],
    "2": % DeviceRetrievalMethods (Device engagement using QR code with BLE for retrieval)
    [
      [
        2, % BLE
        1, % Version
        { % BLE options
          "0": false, % no support for mdoc peripheral server mode
          "1": true,  % support for mdoc central client mode
          "11": "45EFEF742B2C4837A9A3B0E1D05A6917" % UUID of mdoc client central mode
        }
      ]
    ],
    "5": % OriginInfos (Required because Capabilities is present)
    [],
    "6": % Capabilities (Defines supported features)
    {
      "2": false, % HandoverSessionEstablishmentSupport (Supports negotiated handover)
      "3": true % ReaderAuthAllSupport (Supports reader authentication)
    }
  }

**Step 7**: The verifier uses its Relying Party Instance to scan the QR code and retrieve the DeviceEngagement data from the mdoc.

> **Step 8**: The Relying Party Instance generates its ephemeral key pair (``EReaderKey. Priv``, ``EReaderKey.Pub``). The private key (``EReaderKey.Priv``) MUST be kept secret, and the public key (``EReaderKey.Pub``) MUST be used in establishing the session.

Reviewer: ```suggestion
**Step 8**: The Relying Party Instance generates its ephemeral key pair (``EReaderKey.Priv``, ``EReaderKey.Pub``). The private key (``EReaderKey.Priv``) MUST be kept secret, and the public key (``EReaderKey.Pub``) MUST be used in establishing the session.
```

                            "docType": "org.iso.18013.5.1.it.WalletAttestation",
                            "valueDigests": {
                                "org.iso.18013.5.1.it": {
                                    "0": "h751ADCEBF",
                                    "1": "hE585394618",
                                    "2": "h6345394573",
                                    "3": "h67F5394481",
                                },
                                "org.iso.18013.5.1": {"0": "h751A5FEBF"},
                            },
                            "deviceKeyInfo": {
                                "deviceKey": {"1": 2, "-1": 1, "-2": "h963D1F9A", "-3": "h1FBDC3D6"},
                            },
                            "validityInfo": {
                                "signed": "2020-10-01T13:30:02Z",
                                "validFrom": "2020-10-01T13:30:02Z",
                                "validUntil": "2021-10-01T13:30:02Z",
                            },
                        },
                    },
                    "h59E64205DF1E766AEFF13CB2E",
                ],
            },
        },
    ],
  }

> **Step 17**: The Relying Party Instance receives the SessionData, then it MUST decrypt it, and it MUST verify the Wallet Instance's signature to ensure the data's integrity and that it originates from the expected device (device binding). It also MUST check the validity of the mdoc, including its issuer's signature. In case of long-lived Digital Credential it SHOULD also check the revocation status.

Reviewer: ```suggestion
**Step 17**: The Relying Party Instance receives the SessionData, then it MUST decrypt it, and it MUST verify the Wallet Instance's signature to ensure the data's integrity and that it originates from the expected device (device binding). It also MUST check the validity of the mdoc, including its issuer's signature. In case of long-lived Digital Credentials, it SHOULD also check the revocation status using :ref:`OAuth Status List`.
```

Relying Party and Wallet Instances registered in the IT-Wallet ecosystem MUST support at least:

- *Supervised Device Retrieval flow* where a human Relying Party is overseeing the verification process in person, in contrast with *unsupervised flow* where verification might happen through automated systems without human oversight.
- *Device Engagement* based on QR Code.
- *RP Instance Authentication* following the mechanisms defined in the `ISO18013-5`_ for the *reader authentication*.
- *Device Retrieval* mechanism based on Bluetooth Low Energy (BLE) for the communication sub-phase. *Server Retrieval* mechanism MUST NOT be supported.
- Domestic *Document Type* and *Namespaces* defined in this technical specification in addition to those already defined in the `ISO18013-5`_ for the mDL (see Section :ref:`MDOC-CBOR` for more details).
- *Wallet Instance validation* through the Wallet Attestation.

The following figure illustrates the low-level flow compliant with ISO 18013-5 for proximity flow.

.. _fig_High-Level-Flow-ITWallet-Presentation-ISO-updated:
.. figure:: ../../images/High-Level-Flow-ITWallet-Presentation-ISO-updated.svg
    :figwidth: 100%
    :align: center

>     :target: https://www.plantuml.com/plantuml/svg/ZLF1RXD13BtdAwmzKQdXmB52LQ694094IQXwuUBCECcekvx1uyrg_nxlxZHAH8FOGwJCFB_t_EovougYRLFFiDN8RRCXcSt0FdGQ1Hup2M26TZmRVv8L11SIiaBrX7LD2eySj1fmtT-GrGDyPfDXHoFaCF6ty_djRNS3tnCnw9vWeZfY9z3xkM6AYzPGRBoe8J9i-jUg8CfmW_2qlb-y2DqUEJfpb44Z9A5iPD3uw60HTa6nXdiXFoXWdSVAc1I-8NlKA3s4BQoLqlmTX0mS5HB9Daf3tXGdFR2lCyljHSj76B6Vu9ExJyGag0HC7L3QKqDYZfwfXuH1u1U_gQhbWWx1qHVgtpv8E5JLDPmVjfkgkYgqndkuWVK-TcDEFrPm7pr1O5UtFl3EIgVA9UzmPtgiLJNM7ifarWxvjBPS6k__-8N1y8hGavmRBeVdXWklwJIfLO9CEOV80s7PlFATWX-44We1GWux9W_87eYTz4c7VweIDiPcReH-jvHrCBMUg9TPSLE7l9ywBowcpY-181v8uEFN9MHohH2uFJ9JVJh9InsKwmMDzYvB33sOIjZs43IFdQ6Qrf74bcjLki6K91wm1jl3bvKUHfkf9dSMGRws45na9JRIk0MXNSWgqs8w4WtROKtfzCbyAFIJT5aeOdyfS3EG8cSwINc1YXQkYNT1zr36SERidw6MIP-dTkIzVZMfuR7uSRLX4tZOzQl1gF6IUvvUl3KLVD3--69Dep33Jv8cy1ZCx8xOjqtz1m00

Reviewer 2: ```suggestion
    :target: https://www.plantuml.com/plantuml/svg/ZLF1RXCn4BtdAqPxebBBWME5K4iR8GI8a53rmiMn9udLtR6nnxlaxndR6yaHZSGUejYUUU_Dy_DTP1Ku3Vr29NQKXYs6nGBaKvfTWAU80LI4LFRAZvS6gwrNnD0zgxv7XYUotEMwwT-IzGe-az2Gnb8oLlnqys6wXUyU2VY0C0ETAayIxAWOPzJqlhHKIKcQI4WKMqUmFRu60Yxa4mNFcuVxXhL2HEnHSooDFhW_Th1yb7yO1RZ2xXBf_4VrSbJwdMVufiWeCUS8TWSRHW_lm4OWn-0nx5mOq0XlPGNv9X6UBWXlof1CBIyQFo5XoCCJJe3-W8CiUtX1qQXiW5_qfkkMZpXQZL_m-7OLxDXrrt2-cRPTTtL27MzXSt1JcVZ76XwSGR2a-sGixQMnzYrfh_R2HjnrvzSmbkLuK_HjjA8MgPu9oVz8XwnzZSfWQBMtcuts2YiiWq-26Z76x9tWdL9PlwPs5L12YA0WsZsX0KK6a7GuUiuhvc2t0YPQ0lvE63bBsbJaPMIrxMqyL1Eksh4l5O4RCS37hqk8g9CmwuCYmhwvqKlww-aZ5d6N2V84mc5taZkEPIx5mZHZb8YjvrcKbqIFuhcayHHfkUCvywq1yLeiq8uOIK3T3WBkRUHGuwaniDtbi6BpVEvdz3PNIEQIsKBl7KLi77vKhCHOePui7w1UmRckIuOsbznOri9UlsfJVZ2c11osbdBYhx9EWKNz0eDn6cGaQ-fBhuzME6Pa-8bXi5HSASS6ssTnlZ4jxf1C6lg_
```

    
    High-Level Proximity Flow

**Step 1**: The User opens the Wallet Instance initiating the process.

**Step 2**: The User authenticates itself to the Wallet Instance. This can be done by the Wallet Instance or a Wallet Secure Cryptographic Device (WSCD). It is a prerequisite for accessing sensitive data and presenting attributes.

> **Step 3**: The User explicitly indicates their intention to present their mdoc Digital Credentials.

Reviewer 3: I don't get it. Here the user knows that he has to present something but he doesn't know what yet and he can't even give explicit consent. The credential to present is requested by the verifier.

Reviewer 4: +1, we are going to update the figure and the text with the following sentence: "Step 3: User selects the proximity presentation functionality."

Author: The figure and text are updated accordingly.

> **Step 4**: [Optional] If the initial authentication in Step 2 was not done through WSCD, a separate authentication via WSCD MAY be required.

Reviewer 3: I'm not clear on what WSCD authentication means. I'd call it biometric unlocking

Reviewer 4: Biometric unlocking is an instance of WSCA authentication. Here we were more general as specified in the ARF (Discussion Topic C, req WUA_03). We are going to reword WSCD with WSCA.

Author: The figure and textual description are revised accordingly.

Below is a non-normative example using the diagnostic notation of a CBOR-encoded ``SessionEstablishment`` that contains the mdoc request of a Wallet Attestation along with an mDL Digital Credential.

.. literalinclude:: ../../examples/iso-session-establishment.txt
  :language: text

**Step 11**: The Relying Party Instance MUST transmit the encrypted and signed ``SessionEstablishment`` message to the Wallet Instance over a secure BLE connection that was established based on the device engagement information.

**Step 12**: The Wallet Instance MUST compute the session key, as described in Step 9.

**Step 13**: Upon receiving the ``SessionEstablishment`` message, the Wallet Instance MUST decrypt it using the shared session key and MUST verify the Relying Party Instance's signature (mdoc reader authentication as specified in [`ISO18013-5`_ #9.1.1.4]) to ensure its authenticity.

> **Step 14**: The Wallet Instance MUST decrypt the attribute request and MUST prompt the User for their consent to share the requested attributes. It MUST also display the contents of the Relying Party's registration certificate to ensure transparency about the requested data and its registered purpose.

Reviewer 5: ```suggestion
**Step 14**: The Wallet Instance MUST decrypt the attribute request and MUST prompt the User for their consent to release the requested attributes. It MUST also display the contents of the Relying Party's registration certificate to ensure transparency about the requested data and its registered purpose.
```

do we have the definition or the scope of the registration certificate?

Author: According to the ARF v 1.8, when a Relying Party successfully registers, the Registrar issues a registration certificate to the Relying Party. Section 6.6.3.3 describes the purpose of the registration certificate. 

During registration, the Relying Party specifies the attributes it intends to request from Wallet Units. The Registrar then lists these attributes in the Relying Party registration certificate and sends it to the Relying Party, distributing it to all its Relying Party Instances. 

During a transaction, a Relying Party Instance sends this registration certificate to the Wallet Unit in the presentation request. The Wallet Unit then displays the contents of the registration certificate to the User when asking for approval, particularly if one or more of the requested attributes are not included in the list of attributes in the registration certificate. This process ensures transparency, allowing Users to make informed decisions about whether to approve the presentation of the requested attributes.

As per ARF v1.8, the format of the registration certificate and the method by which the Wallet Unit verifies that it belongs to the authenticated Relying Party WILL BE specified in a technical specification. However, [A.2.3.44 Topic 44—Relying Party registration certificates] identifies high-level requirements for the Issuance, Presentation, and Verification of the Relying Party registration certificates.

       
       This component MUST be present only if `ReaderAuthAllSupport` is set to `true` in the DeviceEngagement structure, and individual `readerAuth` fields are not used.

mdoc Response
-------------

The messages in the mdoc Response MUST be encoded using CBOR and MUST be encrypted with the Session Key obtained after the Device Engagement phase.
Each mdoc Response MUST be compliant with the following structure, and MUST include the following components, unless otherwise specified:

.. _table-mdoc-attributes:

.. list-table::
   :widths: 25 75
   :header-rows: 1

   * - **Component**
     - **Description**

   * - **version**
     - *(tstr)*. Version of the mdoc Response structure. Enables tracking changes and maintaining compatibility across versions of the standard or implementation profiles.

   * - **documents**
     - *(array of Documents, OPTIONAL)*. CBOR-encoded collection of documents returned in response to the request. Each document includes `issuerSigned` and `deviceSigned` components, and follows the structure defined in the below table.

   * - **documentErrors**

>      - *(map, OPTIONAL)*. A map of error codes for unreturned documents, as defined in [`ISO18013-5`_ #8.3.2.1.2.3]. Each key is a `docType`, and each value is an `ErrorCode` (int) indicating why the document was not returned.

Reviewer 5: ```suggestion
     - *(map, OPTIONAL)*. A map of error codes for unreturned documents, as defined in [`ISO18013-5`_ #8.3.2.1.2.3]. Each key is a `docType`, and each value is an `ErrorCode` (int) indicating the reason why the document was not returned.
```

   * - **status**
     - *(uint)*. Status code indicating the outcome of the request. For example, `"status": 0` means successful processing. For details, see Table 8 (ResponseStatus) of [`ISO18013-5`_ #8.3.2.1.2].

> Each document in **documents** MUST be compliant with the following structure, and MUST include the following components, unless otherwise specified:

Reviewer 5: ```suggestion
Each document in **documents** MUST be compliant with the following structure, and it MUST include the following components, unless otherwise specified:
```

Proximity Flow
==============

> This section describes how a Relying Party requests the presentation of an *mdoc-CBOR* Credential to a Wallet Unit according to the *ISO 18013-5 Specification*. Only *Supervised Device Retrieval flow* is supported in this technical implementation profile. 

Reviewer 5: RP is also known as Credential Verifier

The following figure illustrates the flow diagram compliant with ISO 18013-5 for proximity flow.

.. _fig_High-Level-Flow-ITWallet-Presentation-ISO-updated:
.. figure:: ../../images/High-Level-Flow-ITWallet-Presentation-ISO-updated.svg
    :figwidth: 100%
    :align: center
    :target: https://www.plantuml.com/plantuml/svg/ZLF1RXD13BtdAwmzKQdXmB52LQ694094IQXwuUBCECcekvx1uyrg_nxlxZHAH8FOGwJCFB_t_EovougYRLFFiDN8RRCXcSt0FdGQ1Hup2M26TZmRVv8L11SIiaBrX7LD2eySj1fmtT-GrGDyPfDXHoFaCF6ty_djRNS3tnCnw9vWeZfY9z3xkM6AYzPGRBoe8J9i-jUg8CfmW_2qlb-y2DqUEJfpb44Z9A5iPD3uw60HTa6nXdiXFoXWdSVAc1I-8NlKA3s4BQoLqlmTX0mS5HB9Daf3tXGdFR2lCyljHSj76B6Vu9ExJyGag0HC7L3QKqDYZfwfXuH1u1U_gQhbWWx1qHVgtpv8E5JLDPmVjfkgkYgqndkuWVK-TcDEFrPm7pr1O5UtFl3EIgVA9UzmPtgiLJNM7ifarWxvjBPS6k__-8N1y8hGavmRBeVdXWklwJIfLO9CEOV80s7PlFATWX-44We1GWux9W_87eYTz4c7VweIDiPcReH-jvHrCBMUg9TPSLE7l9ywBowcpY-181v8uEFN9MHohH2uFJ9JVJh9InsKwmMDzYvB33sOIjZs43IFdQ6Qrf74bcjLki6K91wm1jl3bvKUHfkf9dSMGRws45na9JRIk0MXNSWgqs8w4WtROKtfzCbyAFIJT5aeOdyfS3EG8cSwINc1YXQkYNT1zr36SERidw6MIP-dTkIzVZMfuR7uSRLX4tZOzQl1gF6IUvvUl3KLVD3--69Dep33Jv8cy1ZCx8xOjqtz1m00
    
    High-Level Proximity Flow

> **Step 1**: The User opens the Wallet Instance. The user initiates the process by opening the Wallet Instance on their device.

Reviewer 5: ```suggestion
**Step 1**: The User opens the Wallet Instance. The User initiates the process by opening the Wallet Instance on their device.
```

**Step 3**: The User explicitly indicates their intention to present their mdoc digital credentials, which is considered an attestation within the Wallet Unit.

**Step 4**: [Optional] If the initial authentication in Step 2 was not done through WSCD, a separate authentication via WSCD might be required before presenting sensitive credentials like the mdoc.

**Step 5**: The Wallet Unit generates a new ephemeral Elliptic Curve key pair for secure communication. The public key (EDeviceKey.Pub) will be used for session encryption. This is part of the device engagement process.

**Step 6**: The Wallet Unit presents a QR Code to the Relying Party Instance. This QR code contains the DeviceEngagement data, which includes the EDeviceKey.Pub and information about supported copter suits. The QR code is one of the defined transmission technologies for device engagement.

Below an example of a device engagement structure that utilizes QR for device engagement and Bluetooth Low Energy (BLE) for data retrieval.

> CBOR data:

Reviewer 5: ```suggestion
CBOR data represented in AF Binary format:
```

> In diagnostic notation:

Reviewer 5: ```suggestion
The same content shown above using the diagnostic notation:
```

**Step 8**: The Relying Party Instance also generates its ephemeral key pair (EReaderKey. Priv, EReaderKey.Pub). The private key (EReaderKey.Priv) is kept secret, and the public key (EReaderKey.Pub) will be used in establishing the session.

**Step 9**: In essence, the Wallet Unit and Relying Party Instance independently derive the session keys using their private ephemeral key and the other party’s public ephemeral key through a key agreement protocol. This ensures session encryption. In this particular step, the Relying Party Instance computes its session key.

**Step 10**: The RP Instance prepares a SessionEstablishment message. This message is signed by the Relying Party Instance (mdoc reader authentication as specified in [`ISO18013-5`_ #9.1.1.4]) and encrypted using the session keys derived in the previous step. The SessionEstablishment message includes the EReaderKey.Pub and a request for specific attribute(s).

The doc request (Wallet Attestation (WA) and an mDL have been requested) MUST be encoded in CBOR, as demonstrated in the following non-normative example.

> CBOR data: 

Reviewer 5: ```suggestion
CBOR data in AF Binary format: 
```

                                    "1": "h67E5394571",
                                    "2": "h33943BF5AE",
                                    "3": "h2E35613D55",
                                    "4": "hEA5C330D59",
                                    "5": "hFAE487AA1D",
                                    "6": "h7D83E7533",
                                    "7": "hF05498936",
                                    "8": "hB68C87F66",
                                    "9": "h0B35D6E0C",
                                    "10": "hC98A14881",
                                    "11": "hB57DDA948",
                                    "12": "h651F812DA",
                                },
                            },
                            "deviceKeyInfo": {
                                "deviceKey": {"1": 2, "-1": 1, "-2": "h9631F9A", "-3": "h1FB3C3D6"},
                            },
                            "validityInfo": {
                                "signed": "2020-10-01T13:30:02Z",
                                "validFrom": "2020-10-01T13:30:02Z",
                                "validUntil": "2021-10-01T13:30:02Z",

>                             },

Reviewer: ```suggestion
                            },
                            "status": {
                                "status_list" : {
                                 "bits": 1,
                                 "lst": "eNrbuRgAAhcBXQ"                             
                                 }
                            }
```

                    },
                    "h59E64205DF1E766AEFF13CB2E",
                ],
                "deviceSigned": {
                    "nameSpaces": {},
                    "deviceAuth": {
                        "deviceSig": [
                            {"1": 5},
                            {},
                            null,
                            "hE99521A85AD75943BF5AEAE68943BF5AE249943BF5AE943BF5AE43BF5AED",
                        ],
                    },
                },
            },
        },
        {
            "docType": "org.iso.18013.5.1.it.WalletAttestation",
            "issuerSigned": {
                "nameSpaces": {
                    "org.iso.18013.5.1": [
                        {
                            "digestID": 0,
                            "random": "h37986C20A9",
                            "elementIdentifier": "issuing_authority",
                            "elementValue": "ABC123XYZ",
                        },
                    ],
                    "org.iso.18013.5.1.it": [
                        {
                            "digestID": 0,
                            "random": "h37986C20A9",

>                             "elementIdentifier": "wallet_id",

Reviewer: ```suggestion
                            "elementIdentifier": "wallet_name",
```

**Step 18**: Once the data exchange is complete, either party can terminate the session. If BLE is used, this can involve sending a status code for session termination or the “End” command. In this scenario, the GATT Client (Relying Party Instance) MUST unsubscribe from characteristics and disconnect from the GATT server (Wallet Instance). 

**Final Consideration**: The presentation flow focused on the technical data exchange in proximity settings. It is crucial to recognise that supervised proximity flows involving a human verifier play a vital role in many use cases (e.g., age verification at a store, identity check by law enforcement). The human element adds a layer of identity verification through visual inspection and comparison, contributing to User Binding and overall authentication assurance aspects not fully captured in a purely technical presentation flow.

> 

Reviewer: ```suggestion

.. note::

  During proximity presentation the Wallet Instance might not be able to fetch a fresh Wallet Attestation, in this case, the Wallet Instance SHOULD send the latest version of the Wallet Attestation. It is left up to the Relying Party to determine whether a presentation with a valid but expired Wallet Attestation is valid or not.
```
